File: apps/windows/TeslaSync.App/feature-views/PedalUsage/PedalUsage.Model.cs
using System.Globalization;
using System.Text.Json;
using TeslaSync.App.Core;
using TeslaSync.App.Core.Data.State;
using TeslaSync.App.Core.Notifications;
using TeslaSync.App.Core.Units;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The mutually-exclusive lifecycle state the <c>PedalUsage</c> surface can be in — the native superset of the
/// branches the web component renders
/// (web/src/features/driving/components/driving-dynamics/PedalUsage.tsx). The web component is a pure child that
/// reads one live snapshot via <c>useDriveDynamicsLatest(vehicleId)</c> and shows either its three-up pedal gauge
/// row (when any pedal signal is present) or a single empty state. The native surface binds its own
/// cache-then-network read of that snapshot, so it owns the full loading / ready / empty / error / stale /
/// offline matrix the P2 state contract requires. Every value maps onto a visible surface — none is hidden
/// behind a <c>{data &amp;&amp; …}</c> guard.
/// </summary>
public enum PedalUsageState
{
    /// <summary>Initial fetch with no cached snapshot — render the skeleton chrome.</summary>
    Loading,

    /// <summary>A fresh snapshot carrying at least one pedal signal — render the three-up gauge row.</summary>
    Ready,

    /// <summary>The snapshot resolved but carried no pedal telemetry (web <c>!hasAny</c>) — render the empty state.</summary>
    Empty,

    /// <summary>The request failed and no cached snapshot exists — render the retry affordance.</summary>
    Error,

    /// <summary>A cached snapshot older than the freshness window — render the gauges plus a stale chip.</summary>
    Stale,

    /// <summary>The network failed but a cached snapshot remains — render the gauges plus an offline chip.</summary>
    Offline,
}

/// <summary>
/// The categorical accent a pedal gauge renders its value arc with — the native mirror of the literal hue the web
/// source passes each <c>RadialGauge</c> (<c>#06b6d4</c> cyan for throttle, <c>#ef4444</c> red for brake). Kept
/// WinUI-free so the projection can assign and the tests can assert the per-gauge colour without a UI host; the
/// view maps each accent to a themed chart brush (whose token colour is the same hex) at render time.
/// </summary>
public enum PedalGaugeAccent
{
    /// <summary>Cyan (web <c>#06b6d4</c>) — the throttle gauge; maps to the themed <c>TsChartRegenBrush</c>.</summary>
    Cyan,

    /// <summary>Red (web <c>#ef4444</c>) — the brake gauge; maps to the themed <c>TsChartTemperatureBrush</c>.</summary>
    Red,
}

/// <summary>
/// The canonical pedal-telemetry inputs the web <c>PedalUsage</c> consumes — the native mirror of the three
/// <c>DriveDynamicsSnapshot</c> fields the web source actually reads (web/src/api/types.ts):
/// <c>pedal_position</c>, <c>brake_pedal_position</c> and <c>brake_pedal_active</c>. Throttle and brake are pedal
/// positions in percent (0..100); brake-active is a boolean. Each is <see langword="null"/> when the signal is
/// absent, exactly as the web narrows <c>typeof … === 'number' / 'boolean' ? … : null</c>. Pure data — no WinUI
/// types — so the parse and the projection are unit-tested without a UI host.
/// </summary>
/// <param name="ThrottlePercent">Throttle pedal position 0..100 (%), or null (web <c>pedal_position</c>).</param>
/// <param name="BrakePercent">Brake pedal position 0..100 (%), or null (web <c>brake_pedal_position</c>).</param>
/// <param name="BrakeActive">Whether the brake pedal is depressed, or null (web <c>brake_pedal_active</c>).</param>
public sealed record PedalReading(
    double? ThrottlePercent,
    double? BrakePercent,
    bool? BrakeActive)
{
    /// <summary>The no-telemetry reading — the parse fallback for an absent / non-object body (web <c>data == null</c>).</summary>
    public static PedalReading Empty { get; } = new(null, null, null);

    /// <summary>
    /// True when any pedal signal is present — the native port of the web
    /// <c>hasAny = throttle != null || brakePos != null || brakeActive != null</c>. The surface shows the gauge
    /// row when this is true and the empty state otherwise.
    /// </summary>
    public bool HasData => ThrottlePercent is not null || BrakePercent is not null || BrakeActive is not null;

    /// <summary>
    /// Project a <c>GET /drive-dynamics/latest</c> JSON object into the pedal inputs — the native port of the
    /// field narrowing the web source performs on the snapshot. A non-object body yields <see cref="Empty"/>;
    /// each field is read null-tolerantly (a missing / wrong-kind field becomes null) so a partial snapshot never
    /// throws. Throttle / brake accept a numeric or numeric-string value (NaN / Infinity rejected); brake-active
    /// accepts only a JSON boolean (web <c>typeof … === 'boolean'</c>).
    /// </summary>
    /// <param name="element">The parsed drive-dynamics snapshot body.</param>
    /// <returns>The pedal inputs, or <see cref="Empty"/> when there is no telemetry.</returns>
    public static PedalReading FromSnapshotJson(JsonElement element)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            return Empty;
        }

        return new PedalReading(
            PedalReadingJson.GetDouble(element, "pedal_position"),
            PedalReadingJson.GetDouble(element, "brake_pedal_position"),
            PedalReadingJson.GetBool(element, "brake_pedal_active"));
    }
}

/// <summary>
/// Null-tolerant <see cref="JsonElement"/> readers for the pedal snapshot — the numeric getter tolerates a
/// numeric or numeric-string field and rejects NaN / Infinity, and the boolean getter accepts only a true JSON
/// boolean, so a partial or schema-drifted snapshot never aborts the parse (web parity: the source narrows each
/// field by <c>typeof</c> and falls back to null). WinUI-free.
/// </summary>
internal static class PedalReadingJson
{
    /// <summary>The numeric value of <paramref name="name"/>, tolerating a numeric or numeric-string field.</summary>
    public static double? GetDouble(JsonElement obj, string name)
    {
        if (obj.ValueKind != JsonValueKind.Object || !obj.TryGetProperty(name, out var prop))
        {
            return null;
        }

        return prop.ValueKind switch
        {
            JsonValueKind.Number when prop.TryGetDouble(out var n) && !double.IsNaN(n) && !double.IsInfinity(n) => n,
            JsonValueKind.String when double.TryParse(prop.GetString(), NumberStyles.Float, CultureInfo.InvariantCulture, out var s) && !double.IsNaN(s) && !double.IsInfinity(s) => s,
            _ => null,
        };
    }

    /// <summary>The boolean value of <paramref name="name"/>, or null when absent or not a JSON boolean.</summary>
    public static bool? GetBool(JsonElement obj, string name)
    {
        if (obj.ValueKind != JsonValueKind.Object || !obj.TryGetProperty(name, out var prop))
        {
            return null;
        }

        return prop.ValueKind switch
        {
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            _ => null,
        };
    }
}

/// <summary>
/// One projected, render-ready pedal gauge — the native analogue of one web <c>&lt;RadialGauge&gt;</c> plus its
/// caption. Holds the gauge's own label (web <c>RadialGauge label</c>), the caption shown beneath it (web
/// <c>&lt;span&gt;</c>), the clamped value and its full-sweep maximum (so the arc sweep matches the web), the
/// literal unit suffix (web <c>'%'</c> when present, an em dash otherwise), the decimal precision (web
/// <c>Number.isInteger(clamped) ? 0 : precision</c>), the categorical accent (so the arc colour matches the web
/// hue) and a Narrator name. Pure data.
/// </summary>
/// <param name="GaugeLabel">Label shown inside the gauge (web <c>RadialGauge label</c>, e.g. "Throttle").</param>
/// <param name="CaptionText">Caption shown beneath the gauge (web caption, e.g. "Throttle Position").</param>
/// <param name="Value">The clamped value the gauge displays (web <c>value</c>, clamped to [0, max]).</param>
/// <param name="Max">The value mapped to a full sweep (web <c>max={100}</c>).</param>
/// <param name="Unit">The literal unit suffix (web <c>'%'</c> or the em dash when the value is absent).</param>
/// <param name="Decimals">Fraction digits for the rendered value.</param>
/// <param name="Accent">The categorical value-arc accent (web <c>RadialGauge color</c>).</param>
/// <param name="AutomationName">The composed Narrator name for the gauge tile.</param>
public sealed record PedalGaugeDisplayItem(
    string GaugeLabel,
    string CaptionText,
    double Value,
    double Max,
    string Unit,
    int Decimals,
    PedalGaugeAccent Accent,
    string AutomationName);

/// <summary>
/// The projected, render-ready brake-pedal status tile — the native analogue of the web Footprints-icon + status
/// <c>&lt;Badge&gt;</c> + caption column. <see cref="BadgeText"/> is the localized active / inactive copy,
/// <see cref="BadgeStatus"/> the semantic colour (web <c>variant: brakeActive ? 'danger' : 'success'</c>) and
/// <see cref="CaptionText"/> the caption beneath. Pure data.
/// </summary>
/// <param name="BadgeText">Localized brake status (web "Brake Active" / "Brake Inactive").</param>
/// <param name="BadgeStatus">Semantic badge colour: danger when active, success otherwise.</param>
/// <param name="CaptionText">Caption beneath the badge (web "Brake Pedal Status").</param>
/// <param name="AutomationName">The composed Narrator name for the status tile.</param>
public sealed record PedalBrakeStatusContent(
    string BadgeText,
    StatusKind BadgeStatus,
    string CaptionText,
    string AutomationName);

/// <summary>
/// The fully projected, render-ready content of the <c>PedalUsage</c> surface — the native analogue of everything
/// the web component composes before returning its panel. Holds the localized panel <see cref="Title"/>, the data
/// flag (web <c>hasAny</c>), the throttle and brake gauges, the brake-status tile, the empty-state message and the
/// surface's accessible name. The per-state chrome (loading / error / stale / offline) is layered by the
/// view-model + view; this record carries the data-bearing content. Pure data so the projection is unit-tested
/// without a UI host.
/// </summary>
/// <param name="Title">The localized panel title (web "Pedal Usage").</param>
/// <param name="HasData">True when any pedal signal is present (web <c>hasAny</c>).</param>
/// <param name="Throttle">The projected throttle gauge.</param>
/// <param name="Brake">The projected brake gauge.</param>
/// <param name="BrakeStatus">The projected brake-status tile.</param>
/// <param name="EmptyMessage">The empty-state message (web "No pedal telemetry received yet").</param>
/// <param name="AutomationName">The composed Narrator name for the populated surface.</param>
public sealed record PedalUsageContent(
    string Title,
    bool HasData,
    PedalGaugeDisplayItem Throttle,
    PedalGaugeDisplayItem Brake,
    PedalBrakeStatusContent BrakeStatus,
    string EmptyMessage,
    string AutomationName);

/// <summary>
/// Pure projection from a <see cref="PedalReading"/> to its render-ready <see cref="PedalUsageContent"/> — the
/// native port of the render logic in web/src/features/driving/components/driving-dynamics/PedalUsage.tsx. The
/// gauges reproduce the web call sites one-for-one: each value is clamped to [0, 100] exactly as the web
/// RadialGauge, the unit is <c>'%'</c> when the signal is present and an em dash otherwise (web
/// <c>unit={x != null ? '%' : '—'}</c>), the precision is the web integer-or-global-precision rule, and the brake
/// status maps <c>brakeActive ? danger/'Brake Active' : success/'Brake Inactive'</c>. Every translatable label
/// resolves through the i18n facade using the same keys the web source passes to <c>t()</c>. No WinUI types —
/// unit-tested without a UI host.
/// </summary>
public static class PedalUsageProjection
{
    /// <summary>The full-sweep maximum for both pedal gauges (web <c>max={100}</c>).</summary>
    public const double GaugeMax = 100;

    /// <summary>The unit suffix shown when a pedal value is present (web <c>'%'</c>).</summary>
    public const string PercentUnit = "%";

    /// <summary>The unit suffix shown when a pedal value is absent (web em dash <c>'—'</c>).</summary>
    public const string UnknownUnit = "\u2014";

    /// <summary>The global display precision the web RadialGauge uses for a non-integer value (web <c>getGlobalPrecision()</c>).</summary>
    public const int DisplayPrecision = 2;

    /// <summary>Project <paramref name="reading"/> into the render-ready content using the i18n facade.</summary>
    /// <param name="reading">The pedal-telemetry inputs.</param>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    /// <returns>The display-ready content (gauges, brake status, copy and accessible name).</returns>
    public static PedalUsageContent Project(PedalReading reading, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(reading);
        ArgumentNullException.ThrowIfNull(localizer);

        string title = localizer.GetString("dynamics.pedalUsage", "Pedal Usage");
        string emptyMessage = localizer.GetString("dynamics.pedalNoData", "No pedal telemetry received yet");

        PedalGaugeDisplayItem throttle = BuildGauge(
            localizer.GetString("dynamics.throttle", "Throttle"),
            localizer.GetString("dynamics.throttlePosition", "Throttle Position"),
            reading.ThrottlePercent,
            PedalGaugeAccent.Cyan);

        PedalGaugeDisplayItem brake = BuildGauge(
            localizer.GetString("dynamics.brake", "Brake"),
            localizer.GetString("dynamics.brakePedalPosition", "Brake Pedal Position"),
            reading.BrakePercent,
            PedalGaugeAccent.Red);

        PedalBrakeStatusContent brakeStatus = BuildBrakeStatus(reading.BrakeActive, localizer);

        string automationName = string.Join(
            ". ", title, throttle.AutomationName, brake.AutomationName, brakeStatus.AutomationName);

        return new PedalUsageContent(
            Title: title,
            HasData: reading.HasData,
            Throttle: throttle,
            Brake: brake,
            BrakeStatus: brakeStatus,
            EmptyMessage: emptyMessage,
            AutomationName: automationName);
    }

    /// <summary>
    /// Maps the brake-active flag to its semantic badge colour — the native port of the web
    /// <c>variant: brakeActive ? 'danger' : 'success'</c>. A null (unknown) flag is treated as not-active, exactly
    /// as the web falsy check renders the success "Brake Inactive" badge.
    /// </summary>
    /// <param name="brakeActive">The brake-active flag, or null when unknown.</param>
    /// <returns><see cref="StatusKind.Danger"/> when active, otherwise <see cref="StatusKind.Success"/>.</returns>
    public static StatusKind BrakeStatusFor(bool? brakeActive) =>
        brakeActive == true ? StatusKind.Danger : StatusKind.Success;

    private static PedalGaugeDisplayItem BuildGauge(
        string gaugeLabel,
        string captionText,
        double? percent,
        PedalGaugeAccent accent)
    {
        // web RadialGauge: value={x ?? 0}, clamped internally to [0, max]; unit '%' when present else em dash.
        double value = Math.Clamp(percent ?? 0, 0, GaugeMax);
        string unit = percent is not null ? PercentUnit : UnknownUnit;
        int decimals = DecimalsFor(value);
        string valueText = ScalarFormatters.FormatNumber(value, decimals);

        string automationName = string.Format(
            CultureInfo.CurrentCulture, "{0}, {1}{2}, {3}", gaugeLabel, valueText, unit, captionText);

        return new PedalGaugeDisplayItem(
            gaugeLabel, captionText, value, GaugeMax, unit, decimals, accent, automationName);
    }

    private static PedalBrakeStatusContent BuildBrakeStatus(bool? brakeActive, ILocalizer localizer)
    {
        bool active = brakeActive == true;
        string badgeText = active
            ? localizer.GetString("dynamics.brakeActive", "Brake Active")
            : localizer.GetString("dynamics.brakeInactive", "Brake Inactive");
        string caption = localizer.GetString("dynamics.brakePedal", "Brake Pedal Status");
        StatusKind status = BrakeStatusFor(brakeActive);

        string automationName = string.Format(CultureInfo.CurrentCulture, "{0}, {1}", badgeText, caption);

        return new PedalBrakeStatusContent(badgeText, status, caption, automationName);
    }

    // web RadialGauge: d = Number.isInteger(clamped) ? 0 : getGlobalPrecision(). The web passes no explicit
    // decimals, so an integer value shows none and a fractional one shows the global precision.
    private static int DecimalsFor(double clamped) =>
        clamped == Math.Floor(clamped) ? 0 : DisplayPrecision;
}

/// <summary>
/// Maps the engine's raw <c>RepositoryResult&lt;JsonElement&gt;</c> emissions onto projected
/// <c>RepositoryResult&lt;PedalReading&gt;</c>, preserving every freshness flag (cached / refreshing / stale /
/// offline) so the view-model can render the full state matrix. Pure so the parse-and-preserve contract is
/// unit-tested without a network or cache.
/// </summary>
public static class PedalUsageResultMapper
{
    /// <summary>Project <paramref name="raw"/>'s payload (when present) while preserving its status.</summary>
    /// <param name="raw">The raw cache-then-network emission.</param>
    /// <returns>The same emission with its snapshot body parsed into a <see cref="PedalReading"/>.</returns>
    public static RepositoryResult<PedalReading> Map(RepositoryResult<JsonElement> raw)
    {
        ArgumentNullException.ThrowIfNull(raw);

        PedalReading Parse() => raw.HasValue ? PedalReading.FromSnapshotJson(raw.Value) : PedalReading.Empty;

        return raw.Status switch
        {
            LoadStatus.Loading => RepositoryResult<PedalReading>.Loading(),
            LoadStatus.Cached => RepositoryResult<PedalReading>.Cached(Parse(), raw.FetchedAt!.Value, raw.IsStale),
            LoadStatus.Refreshing => RepositoryResult<PedalReading>.Refreshing(Parse(), raw.FetchedAt!.Value, raw.IsStale),
            LoadStatus.Loaded => RepositoryResult<PedalReading>.Loaded(Parse(), raw.FetchedAt ?? DateTimeOffset.UtcNow),
            LoadStatus.Empty => RepositoryResult<PedalReading>.Empty(raw.FetchedAt),
            LoadStatus.Offline => RepositoryResult<PedalReading>.OfflineCached(Parse(), raw.FetchedAt!.Value, raw.Error!),
            _ => RepositoryResult<PedalReading>.Failure(
                raw.Error ?? new RepositoryError(RepositoryErrorKind.Unknown, "Unknown error")),
        };
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/PedalUsage/PedalUsageModel.kt
// Pure, framework-free model + projection for the PedalUsage feature view — the native analogue of
// everything the web component derives before returning JSX
// (web/src/features/driving/components/driving-dynamics/PedalUsage.tsx). No Compose, no Android, no HTTP:
// every declaration here is unit-tested off-device in the :app:testReleaseUnitTest gate, keeping the
// composable a thin render layer.
//
// PedalUsage is a presentational surface. The web component reads its snapshot from
// `useDriveDynamicsLatest(vehicleId)` (the `/drive-dynamics/latest` projection of the live PedalPosition /
// BrakePedalPos / BrakePedal signals), but — exactly like the sibling LiveMotorStatus / LiveVehicleState
// ports — the native surface takes that decoded snapshot as a prop from its owning Driving Dynamics page,
// which owns the query's loading / error / stale / offline handling. So this surface binds no data fetch;
// its two web data sources are `useTranslation` (the i18n catalog, P1/S10) and `useUnits` (the live decimal
// precision, P1/S8), and the cache-then-network lifecycle states live on the owning page, not here. The two
// branches the web source itself defines are the complete render set:
//   • at least one pedal reading present (web `hasAny`) → the three-up gauge row, and
//   • no reading present → a friendly empty state ("No pedal telemetry received yet"), never a blank box
//     (web `<EmptyState/>`), which doubles as the offline-cached-empty surface.
// A skeleton loading branch is offered behind an opt-in `loading` flag the owning page threads while its
// query is first in flight — the same convention the sibling surfaces use — defaulting to the web's
// no-loading contract.
//
// The web reads a `DriveDynamicsSnapshot` (web/src/api/types.ts); [DriveDynamicsLive] mirrors the slice it
// consumes in snake_case (matching the Go JSON tags served verbatim — no camelCaseKeys transform in the
// shared layer), so the projection runs straight off the cached API JSON. `pedal_position` /
// `brake_pedal_position` are already percentages (0..100), so no unit conversion is applied — the only
// display preference the surface honors is the global decimal precision the web `RadialGauge` reads through
// `getGlobalPrecision()` for a non-integer reading (web `d = decimals ?? (isInteger ? 0 : precision)`); the
// locale-aware number rendering itself is delegated to the shared RadialGauge component, as every gauge
// surface does.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/PedalUsage — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path. `MatchingDeclarationName` is suppressed for the co-located
// supporting types.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.pedalusage

import io.teslasync.shared.core.diagnostics.Logger
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlin.math.floor

/** The em-dash the web renders as a gauge's unit when its reading is absent (`'—'`). */
internal const val DASH: String = "\u2014"

/** Web gauge unit for a present pedal reading (`'%'`). */
internal const val PERCENT_UNIT: String = "%"

/** Pedal-position scale — the web `max={100}` (0..100 %). */
internal const val PEDAL_MAX: Double = 100.0

/** Web `getGlobalPrecision()` default — the user's `decimal_precision`, 2 when unset. */
internal const val DEFAULT_DECIMAL_PRECISION: Int = 2

// Raw `/drive-dynamics/latest` (DriveDynamicsSnapshot) document keys read by the surface — snake_case,
// served verbatim by the Go handler (no camelCaseKeys transform in the shared layer), so the native reads
// match the wire contract.
private const val FIELD_PEDAL_POSITION = "pedal_position"
private const val FIELD_BRAKE_PEDAL_POSITION = "brake_pedal_position"
private const val FIELD_BRAKE_PEDAL_ACTIVE = "brake_pedal_active"

/**
 * The slice of `/drive-dynamics/latest` this surface reads — the native mirror of the `DriveDynamicsSnapshot`
 * fields the web `PedalUsage` consumes (web/src/api/types.ts). Field names keep their snake_case wire form so
 * the projection runs directly off the cached API JSON, and every field is nullable because the backend omits
 * a reading whenever the underlying telemetry has not reported (the web reads each behind a `typeof` guard
 * and falls back to the empty state / em-dash).
 *
 * @property pedalPosition throttle pedal position 0..100 (%), or null when not reported (web `pedal_position`).
 * @property brakePedalPosition brake pedal position 0..100 (%), or null when not reported
 *   (web `brake_pedal_position`).
 * @property brakePedalActive whether the brake pedal is depressed, or null when not reported
 *   (web `brake_pedal_active`).
 */
data class DriveDynamicsLive(
    val pedalPosition: Double?,
    val brakePedalPosition: Double?,
    val brakePedalActive: Boolean?,
) {
    companion object {
        /**
         * Decode a `/drive-dynamics/latest` body into a tolerant snapshot, or `null` when the body is absent
         * / not a JSON object — web parity: `data` is `DriveDynamicsSnapshot | null`. Unlike a present-but-empty
         * snapshot rendering grids elsewhere, PedalUsage gates on `hasAny` (at least one of the three readings
         * present), so a decoded snapshot with all three fields null still resolves to the empty state via the
         * projection, exactly like the web `hasAny` check.
         */
        fun fromJson(element: JsonElement?): DriveDynamicsLive? {
            val obj = element as? JsonObject ?: return null
            return DriveDynamicsLive(
                pedalPosition = obj.numberField(FIELD_PEDAL_POSITION),
                brakePedalPosition = obj.numberField(FIELD_BRAKE_PEDAL_POSITION),
                brakePedalActive = obj.boolField(FIELD_BRAKE_PEDAL_ACTIVE),
            )
        }
    }
}

/**
 * The semantic accent each gauge tints with — the native analogue of the web per-gauge hex (`#06b6d4` cyan /
 * `#ef4444` red). The render layer resolves each to a design token so no hex literal leaks into the view.
 */
enum class PedalAccent { Cyan, Red }

/** Stable identity of each gauge, in the order the web grid emits them. */
enum class PedalGaugeKey { Throttle, Brake }

/**
 * One projected radial gauge — everything the web passes to its `RadialGauge` for the throttle / brake cell.
 * Pure data so the projection is unit-tested without a UI host; the view maps [key] onto its i18n label and
 * [accent] onto a design token, then hands [value] / [max] / [unit] / [decimals] to the shared RadialGauge.
 *
 * @property key the gauge identity (drives the label in the view).
 * @property value the pedal position clamped to `0..`[max] (web `Math.max(0, Math.min(value, max))`).
 * @property max the gauge scale (always [PEDAL_MAX]).
 * @property unit `'%'` when the reading is present, the em-dash otherwise (web `value != null ? '%' : '—'`).
 * @property decimals fraction digits the gauge renders (web `isInteger ? 0 : globalPrecision`).
 * @property accent the arc color (web per-gauge hex), resolved to a token by the view.
 * @property present whether the underlying reading was non-null (drives the unit and is asserted by tests).
 */
data class PedalGauge(
    val key: PedalGaugeKey,
    val value: Double,
    val max: Double,
    val unit: String,
    val decimals: Int,
    val accent: PedalAccent,
    val present: Boolean,
)

/**
 * The localized labels this surface resolves once (P1/S10) and hands to the renderer. Keeping the strings
 * injectable lets the stateless content composable be exercised in a UI test without a resources host and
 * keeps the projection free of any English literal.
 */
data class PedalUsageStrings(
    val title: String,
    val throttle: String,
    val throttlePosition: String,
    val brake: String,
    val brakePedalPosition: String,
    val brakeActive: String,
    val brakeInactive: String,
    val brakePedal: String,
    val noData: String,
    val loadingLabel: String,
)

/**
 * The fully projected, render-ready view — everything the web component computes before returning JSX. Pure
 * data (no Compose types) so the projection is unit-tested without a UI host and each per-state instance
 * doubles as the surface's snapshot.
 *
 * @property loading whether the owning query is still in flight; the surface renders skeleton chrome while
 *   true (the opt-in branch the owning page threads; default false is the web's no-loading contract).
 * @property hasData whether at least one pedal reading is present (web `hasAny`); when false the surface
 *   renders the empty state instead of the gauges.
 * @property throttle the throttle gauge spec (always built so the projection is fully testable; rendered only
 *   when [hasData]).
 * @property brake the brake gauge spec (always built; rendered only when [hasData]).
 * @property brakeActive whether the brake pedal is active — true only when the reading is explicitly true, so
 *   a null/absent reading reads as inactive (web `brakeActive ? … : …`, where null is falsy).
 */
data class PedalUsageDisplay(
    val loading: Boolean,
    val hasData: Boolean,
    val throttle: PedalGauge,
    val brake: PedalGauge,
    val brakeActive: Boolean,
)

/**
 * Pure projection from the surface's prop to its render-ready [PedalUsageDisplay] — a 1:1 port of the
 * derivations the web component performs: the per-field `typeof === 'number' | 'boolean'` guards, the
 * `hasAny` presence gate, the `value ?? 0` clamp the web `RadialGauge` applies, the `value != null ? '%' :
 * '—'` unit, and the `brakeActive ? danger/Active : success/Inactive` badge ternary (null → inactive).
 */
object PedalUsageProjection {
    /**
     * Select the render-ready view for the given [dynamics] snapshot and [loading] flag. [precision] is the
     * user's global decimal precision (web `getGlobalPrecision()`), used only for a non-integer reading.
     */
    fun project(
        dynamics: DriveDynamicsLive?,
        loading: Boolean,
        precision: Int,
    ): PedalUsageDisplay {
        val throttleRaw = dynamics?.pedalPosition
        val brakeRaw = dynamics?.brakePedalPosition
        val brakeActiveRaw = dynamics?.brakePedalActive
        val hasAny = throttleRaw != null || brakeRaw != null || brakeActiveRaw != null
        val safePrecision = precision.coerceAtLeast(0)
        return PedalUsageDisplay(
            loading = loading,
            hasData = hasAny,
            throttle = gauge(PedalGaugeKey.Throttle, throttleRaw, PedalAccent.Cyan, safePrecision),
            brake = gauge(PedalGaugeKey.Brake, brakeRaw, PedalAccent.Red, safePrecision),
            brakeActive = brakeActiveRaw == true,
        )
    }

    /** Build one gauge spec — web `value={raw ?? 0}` clamped to `0..max`, with the present/absent unit. */
    private fun gauge(
        key: PedalGaugeKey,
        raw: Double?,
        accent: PedalAccent,
        precision: Int,
    ): PedalGauge {
        val present = raw != null
        val clamped = (raw ?: 0.0).coerceIn(0.0, PEDAL_MAX)
        return PedalGauge(
            key = key,
            value = clamped,
            max = PEDAL_MAX,
            unit = if (present) PERCENT_UNIT else DASH,
            decimals = gaugeDecimals(clamped, precision),
            accent = accent,
            present = present,
        )
    }

    /**
     * The fraction-digit count the web `RadialGauge` derives when no `decimals` prop is passed:
     * `Number.isInteger(clamped) ? 0 : getGlobalPrecision()`.
     */
    internal fun gaugeDecimals(
        clamped: Double,
        precision: Int,
    ): Int = if (clamped.isFinite() && clamped == floor(clamped)) 0 else precision.coerceAtLeast(0)
}

/**
 * Resolve the global decimal [precision] from the user's settings (web `getGlobalPrecision()`), falling back
 * to [DEFAULT_DECIMAL_PRECISION] when unset — the same default the web helper applies.
 */
internal fun resolveDisplayPrecision(precision: Int?): Int = (precision ?: DEFAULT_DECIMAL_PRECISION).coerceAtLeast(0)

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never a pedal
 * reading — so a diagnostics line can never leak fleet telemetry.
 */
object PedalUsageDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = "PedalUsage"

    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}

/** Read a numeric field, or `null` when absent / JSON `null` / not a JSON number (web `typeof === 'number'`). */
private fun JsonObject.numberField(key: String): Double? = (this[key] as? JsonPrimitive)?.let { if (it.isString) null else it.doubleOrNull }

/** Read a boolean field, or `null` when absent / JSON `null` / not a JSON bool (web `typeof === 'boolean'`). */
private fun JsonObject.boolField(key: String): Boolean? = (this[key] as? JsonPrimitive)?.let { if (it.isString) null else it.booleanOrNull }
